projects/components/src/stepper/stepper.component.ts

```diff
@@ -104,26 +107,35 @@ export class StepperComponent implements AfterContentInit {
 
   public steps$?: Observable<StepperTabComponent[]>;
 
+  public constructor(private readonly cdr: ChangeDetectorRef, private readonly subs: SubscriptionLifecycle) {}
+
   public ngAfterContentInit(): void {
     this.steps$ = queryListAndChanges$(this.steps).pipe(map(list => list.toArray()));
+    this.subs.add(
+      merge(
+        this.selectionChange.pipe(map(({ selectedStep }) => selectedStep)),
+        this.steps$.pipe(map(([firstStep]) => firstStep))
+      )
+        .pipe(
+          filter(step => !isNil(step.stepControl)),
+          switchMap(step => step.stepControl!.statusChanges)
+        )
+        .subscribe(() => this.cdr.markForCheck())
+    );
   }
 
-  // TODO: can be made a directive
-  public isNextDisabled(stepper: MatStepper): boolean {
-    const selectedStep = stepper.steps.get(stepper.selectedIndex);
-    const formValid = selectedStep?.stepControl?.valid;
-    const stepCompleted = selectedStep?.completed;
-
-    return this.isLinear && !stepCompleted && !formValid;
-  }
+  public isNextDisabled(stepIndex: number): boolean {
+    const currentTab = this.steps?.get(stepIndex);
+    if (!currentTab) {
+      return true;
+    }
 
-  // TODO: can be made a directive
-  public isSubmitDisabled(stepper: MatStepper): boolean {
-    const selectedStep = stepper.steps.get(stepper.selectedIndex);
-    const formValid = selectedStep?.stepControl?.valid;
-    const stepCompleted = selectedStep?.completed;
+    const isValid = currentTab.stepControl ? currentTab.stepControl?.status === 'VALID' : currentTab?.completed;
```

Comment:
Looks like we already know currentTab is defined (and stepControl in the true branch)
```suggestion
    const isValid = currentTab.stepControl ? currentTab.stepControl.status === 'VALID' : currentTab.completed;
```